### ISSUE TITLE
Code size measurements in CI

### ISSUE THREAD

**Issue by daverodgman** [ARMmbed/mbedtls#4780]
We would like to track code-size to enable us to understand what factors affect code-size, whether the library is growing, etc.

Measurements:
- library size in kB
- (for PRs) difference for each variant c.f. most recent build of development

Variants:
- platforms should be Arm v7-A and Arm v8-A
- a few different build types, probably: all-options-enabled, one or two other more typical configurations
- static, maybe dynamic?
- Release builds only
- Ubuntu only - skip OS X / Windows / FreeBSD

Reporting:
This should be done via the CI logs - we don't need to aggregate statistics at this time. The difference statistics for a PR should be posted back to the PR in GitHub (example of one way to present this below):

| Variant | Change | Change % | Size with PR |
| ------------- | ------------- | ------------- |------------- |
| Arm v7-A libmbedtls.a all-enabled | 1.1 kB | 5% | 250.5 kB |
| Arm v8-A libmbedx509.so all-enabled | -1.5 kB | -4% | 150.2 kB |

I'm not sure the best way to capture the difference (as this requires knowing the size of the base commit), but I think @bensze01 had some suggestions on this.

**Comment by daverodgman** [comment]
It may also be useful to get data from different compilers.

**Comment by gilles-peskine-arm** [comment]
This is a very broad task. I strongly recommend having a plan written down before starting coding. The people who know the CI best are probably Bence and me.

The CI logs already have library size measurements for baremetal configurations on several Arm platforms. In `all.sh`, see `component_build_armcc` (Arm Compiler 5 and 6) and `component_build_arm_none_eabi_gcc`, `component_build_arm_none_eabi_gcc_arm5vte`, `component_build_arm_none_eabi_gcc_m0plus` (GCC). All of these run on Jenkins, and Travis also runs `build_arm_none_eabi_gcc_arm5vte` and `build_arm_none_eabi_gcc_m0plus`. More similar jobs can be added to have information about different compilers, targets or configurations.

Obviously, if you want to have data from other compilers, they need to be added to the dockerfile.

Comparing with the PR base in a PR job is somewhat complicated. The CI runs on a given commit, not always in the context of a pull request, so there isn't always a base to compare with. The pr-merge job does run extra code which compares a PR's head with a PR's base: see `
gen_abi_api_checking_job` in the Jenkins scripts, which invokes `scripts/abi_check.py` from the Mbed TLS repository. However, this job runs in parallel with the build jobs, so it doesn't have access to their output. Some possible approaches:
* Make the size comparison job re-run the builds on the PR head. Not an attractive solution. The CI is already slow and resource-intensive as it is.
* Make the size measuring jobs additionally measure the size of the PR head if there is a PR head. This would require some pretty messy changes to `all.sh` or to the code that invokes it.
* Make the size comparison job run in a subsequent step, in parallel with `analyze_results`, or as part of it. This means that the size data from head and the size data from base must be available. Doing the builds of base in the second step is not an attractive solution because it will take a long time, so they should be done in the first step. To pass the size data from the first step to the second step, use build artifacts. This is what we already do with [outcome files](https://github.com/ARMmbed/mbedtls/blob/development/docs/architecture/testing/test-framework.md#user-content-outcome-file). I propose to use a similar format for size data: a semicolon-separated CSV where the columns are {platform, configuration, file, text size, rodata size, data size}.

The workflow would thus be:
1. Size measurement jobs in `all.sh` write the size data to `$MBEDTLS_TEST_SIZES_FILE`.
1. The Groovy code runs `all.sh` components with `MBEDTLS_TEST_SIZES_FILE` set to `xxx-sizes.csv`, and uploads the resulting file as an artifact if present.
1. In `run_pr_job`, for a merge job, additionally run the size measurement jobs from `all.sh` for the base.
1. Extend `gather_outcomes` to also assemble `base-sizes.csv` and `head-sizes.csv` and run `tests/scripts/analyze_sizes.py` if present.

Reporting results to GitHub is a completely separate issue and I'm not sure I want it: it's a bit noisy to post a comment on every build. Our pull requests can run into hundreds of review comments and GitHub doesn't scale well. If we do start posting things on GitHub, there's a more important thing that we still haven't done, which is to [post CI results](https://github.com/ARMmbed/mbedtls/issues/3633).

**Comment by gilles-peskine-arm** [comment]
All the parts that are likely to change often (for example, list of configurations, choice of commands to run, result collation…) should be in Python scripts (or modifications of existing shell or perl scripts) in the `mbedtls` repository. Avoid putting more than strictly necessary in the Groovy scripts: they can't be modified or reviewed by people who don't have access to the CI repository, they're hard to test, fewer people can maintain them adequately.

**Comment by daverodgman** [comment]
Re posting CI results: this will be covered by the OpenCI work Linaro are doing.

**Comment by mpg** [comment]
Note: the `baremetal` branch has a [measurement script](https://github.com/ARMmbed/mbedtls/blob/baremetal/scripts/baremetal.sh) which may or may not be useful as an inspiration - it used a specific configuration and platform that was the only target for that branch, which for the present task we probably want to explore different options. Though as Gilles said, this is a very broad task, so perhaps starting with a single config and platform would be a good intermediate step.

**Comment by gilles-peskine-arm** [comment]
For information, we're currently measuring code size in the `baremetal` configuration, but this isn't quite right because it includes debugging features. https://github.com/ARMmbed/mbedtls/pull/4907 creates a new preset configuration `baremetal_size` without debugging features.
